feat: new Kosli Capture drop-in AWS snapshotting tutorial (#337)
kosli-dev/docs: tutorials/snapshotting_with_kosli_capture.mdx

---
title: "Snapshotting AWS runtimes with Kosli Capture"
description: "Drop a single CloudFormation stack into an AWS account and have Kosli Capture discover ECS, EKS, Lambda, and tagged S3 buckets, auto-create Kosli environments, and snapshot them on a schedule."
---

<Info>

> **Kosli Capture is experimental.** The Lambda, CloudFormation template, config schema, and CLI flags it wraps may still change. Reach out to the Kosli team before running it in production, or with any questions and comments.

Reviewer: "the ... CLI flags it wraps" is engineering-insider phrasing, and "questions and comments" reads like soliciting internal review feedback rather than offering customer support. Suggest (swap in a concrete support channel/link if one exists):

```suggestion
**Kosli Capture is experimental.** The Lambda, CloudFormation template, configuration format, and CLI flags may still change between releases. Contact the Kosli team before running it in production — and we'd love to hear your feedback.
```

</Info>

[Kosli Capture](https://github.com/kosli-dev/kosli-capture) is a **single CloudFormation stack** that sits inside your AWS account, discovers the resources you already run, and reports them to Kosli as environment snapshots — with no per-cluster reporter to install, no environments to pre-create, and no inventory list to maintain in config.

> It exists to shorten the runtime-side onboarding story. Getting build and release evidence into Kosli is usually straightforward — you own the pipeline. The runtime side is where most of the friction lives: reporters to deploy into every cluster, environments to create by hand, networking approvals for each workload. Kosli Capture flips that: one stack per account, tags decide what is in scope, and Kosli environments are created on first sight.

Reviewer: Building on the earlier suggestion on this line: I'd go further and drop the vendor-perspective framing entirely — "onboarding story" is internal product vocabulary, and "It exists to..." explains Kosli's motives rather than the reader's problem. Same content, customer's point of view:

```suggestion
Reporting build and release evidence to Kosli is usually straightforward — you own the pipeline. Reporting your runtime environments takes more work: a reporter to deploy into every cluster, environments to create by hand, and networking approvals for each workload. Kosli Capture removes those steps: you deploy one stack per account, tags decide what is in scope, and Kosli environments are created automatically on first sight.
```

- [A Kosli API token](/administration/authentication/service_accounts) with permission to **create environments**, stored as an [SSM SecureString](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html) parameter (default name `/kosli/api-token`).
- Your [Kosli organization name](/administration/authentication/service_accounts).
- **A supported AWS region.** The Kosli CLI Lambda layer is published in nine regions today: `eu-central-1`, `eu-west-1`, `eu-west-2`, `eu-west-3`, `eu-north-1`, `us-east-1`, `us-east-2`, `us-west-1`, `us-west-2`. Deploy the stack into one of these.

You do **not** need to pre-create Kosli environments — Kosli Capture creates them on first sight from the naming template.

## How Kosli Capture works

Kosli Capture is a scheduled Lambda that runs every two minutes (default) inside your AWS account. Each tick it re-derives the full picture from scratch — no state is carried between runs — so a dropped or throttled tick is simply superseded by the next one.

```
resource discovered ──tag filter──▶ env name from template ──▶ ensure env ──▶ snapshot
```

> The Lambda uses the `kosli` CLI from Kosli's **public `kosli-cli` Lambda layer** (`/opt/kosli`) to talk to Kosli, so there is nothing to install into your resources themselves and no CLI binary to vendor into the deploy artefact.

Reviewer: "no CLI binary to vendor into the deploy artefact" justifies a build decision to an internal reviewer — the customer only needs the first half of the sentence:

```suggestion
The Lambda uses the `kosli` CLI from Kosli's **public `kosli-cli` Lambda layer** (`/opt/kosli`) to talk to Kosli, so there is nothing to install into your resources themselves.
```

Two things follow from this design:

- **Config is policy, not inventory.** `config.yaml` never names individual resources. It sets tag filters, naming templates, and per-type toggles. Adding a new cluster does not need a config change.
- **Failures are localized.** A single failing resource never stops the rest of the run — the failure is logged and reported as a `gap`, and the other snapshots complete.

## Deploy Kosli Capture

Clone the repository and deploy from its root:

```bash
git clone https://github.com/kosli-dev/kosli-capture.git
cd kosli-capture
```

Store your Kosli API token in SSM as a SecureString parameter (default name `/kosli/api-token`):

```bash
aws ssm put-parameter \
    --name /kosli/api-token \
    --type SecureString \
    --value "$KOSLI_API_TOKEN"
```

The Makefile drives everything. The first time you deploy into an account/region, run `make bootstrap` to create the artefacts bucket used to upload the Lambda zip:

```bash
make bootstrap REGION=eu-west-1 KOSLI_ORG=<your-org>
```

Then build the function zip, upload it, and deploy the CloudFormation stack:

```bash
make deploy REGION=eu-west-1 KOSLI_ORG=<your-org>
```

> `make deploy` runs `aws cloudformation package` + `deploy`, so **CloudFormation owns the code** — no drift, and rollback reverts the code too. State is server-side (no Terraform, no state bucket).

Reviewer: "State is server-side (no Terraform, no state bucket)" compares against the previous internal iteration of this tool — customers never saw the Terraform version, so the comparison reads as team-facing changelog rather than docs. State the benefit on its own terms:

```suggestion
`make deploy` runs `aws cloudformation package` + `deploy`, so CloudFormation manages both the infrastructure and the function code — a rollback reverts the code too, and there is no separate deployment state for you to manage.
```

<Tip>
Opt into faster deploys with `make deploy DEPLOY_MODE=EXPRESS` (AWS CLI >= 2.35.13; rollback stays on). Run `make help` for the rest of the targets (`clean`, `prune-artefacts`, `destroy-artefacts`, `layer-arn`, `destroy`).
</Tip>

Once the first run completes (every two minutes by default), Kosli environments appear in the [Kosli app](https://app.kosli.com) under **Environments**, named according to the template in `config.yaml`.

## Configure what gets snapshotted

> `config.yaml` at the repo root defines the policy Kosli Capture applies. It has three parts: **naming**, **scope**, and **types**. `make` converts it to `config.json` at build time (stdlib JSON at runtime, no PyYAML), and a malformed file fails the conversion so bad config never ships.

Reviewer: "(stdlib JSON at runtime, no PyYAML)" is an implementation detail aimed at code reviewers — customers don't need to know which parser the Lambda uses. The behavior that matters to them is that a malformed file fails the build:

```suggestion
`config.yaml` at the repo root defines the policy Kosli Capture applies. It has three parts: **naming**, **scope**, and **types**. `make` validates and converts the file at build time, so a malformed config fails the deploy rather than shipping.
```

  lambda:
    enabled: true
    granularity: account           # account = one env per resolved prefix; function = env per function

  s3:
    enabled: true
    include_tags:
      kosli:track: "true"          # S3 is opt-in — see below
```

**S3 is intentionally opt-in.** A Kosli `s3` environment fingerprints bucket **content** (for deployment tracking), not a bucket inventory. Most accounts have many buckets that are not deployment artifacts — Terraform state buckets, CloudFormation staging buckets, log buckets. Snapshotting all of them would be noise. `types.s3.include_tags` must therefore be non-empty; otherwise Kosli Capture records a gap and skips S3 entirely.

### Grouping Lambdas or ECS clusters by tag

Sometimes several Lambda functions or ECS clusters should be reported into the **same** Kosli environment — for example, all the Lambdas that make up one application. Use `grouping` to map an exact env name to a set of tag pairs:

```yaml
types:
  lambda:
    grouping:
      aws-prod-payments: { team: payments }
      aws-prod-checkout: { team: checkout, tier: prod }   # multiple pairs = AND

  ecs:
    grouping:
      aws-prod-platform: { team: platform }
```

> Resources matching **all** pairs of an entry are reported together in one snapshot call (`snapshot lambda --function-names a,b,c` / `snapshot ecs --clusters a,b,c`). Unmatched resources fall through to the default behavior — Lambda's `granularity`, ECS's per-cluster env — unchanged. The first matching entry wins, and an entry with no tags never matches, so a typo cannot sweep in everything.

Reviewer: "so a typo cannot sweep in everything" is the team pre-empting an internal reviewer's objection. State the rules plainly and let them speak for themselves:

```suggestion
Resources matching **all** pairs of an entry are reported together in one snapshot call (`snapshot lambda --function-names a,b,c` / `snapshot ecs --clusters a,b,c`). Unmatched resources fall through to the default behavior — Lambda's `granularity`, ECS's per-cluster env — unchanged. If a resource matches more than one entry, the first entry wins; an entry with no tag pairs never matches anything.
```

> Grouping is only offered where the snapshot command reports a set in one call. S3 (`--bucket` is a single bucket) and per-cluster EKS have no such flag, so grouping does not apply to them — pointing several of those at one env name would just overwrite it each run.

Reviewer: This paragraph defends a design decision ("Grouping is only offered where...", "pointing several of those at one env name would just overwrite it") — that's design-review content, not tutorial content. The customer only needs the rule:

```suggestion
Grouping is available for Lambda and ECS only. S3 buckets and EKS clusters are always reported to their own environments.
```

## Change config without a redeploy

Kosli Capture can read its policy from three places, controlled by the CloudFormation parameter `ConfigSource`:

| Source          | Change flow                                                  | Notes                                                                 |
|-----------------|--------------------------------------------------------------|-----------------------------------------------------------------------|
| `baked` (default) | Edit `config.yaml` → `make deploy`                           | `config.json` is bundled into the Lambda package                     |
| `ssm`           | Edit `config.yaml` → `make seed-config` → next scheduled run | Pushes to the `<name>-config` SSM parameter                          |
| `s3`            | Edit `config.yaml` → `make seed-config` → next scheduled run | Uploads to your `CONFIG_S3_URI` (bring your own bucket)              |

All three deliver the **same** `config.json` produced from `config.yaml`; the only difference is where it lands. `make seed-config` picks the right target from `CONFIG_SOURCE` (and is a no-op for `baked`).

> Runtime sources **fail closed**: a failed fetch reuses the warm sandbox's last-known-good config or fails the run — Kosli Capture never silently falls back to the baked copy, because stale policy could re-include resources you meant to exclude.

Reviewer: "Kosli Capture never silently falls back" is argumentative framing aimed at an internal objection. The fail-closed behavior itself is exactly what customers need — state it as behavior plus benefit:

```suggestion
Runtime sources **fail closed**: if a config fetch fails, Kosli Capture reuses the last-known-good config (when the Lambda sandbox is warm) or fails the run. It does not fall back to the baked copy, so a stale policy can never re-include resources you have excluded.
```

<Tip>
Keep the authoritative `config.yaml` in the repo and push versions via `make seed-config` (or CI) rather than hand-editing the SSM parameter or S3 object.
</Tip>

## EKS caveats — the honest part

> EKS is the one type where a drop-in cannot do everything for you. Two things to be aware of:

Reviewer: "a drop-in" is the team's own vocabulary for the product; a reader who arrived at this page doesn't know it by that name. Also frames it as the product's shortcoming rather than what the reader needs to do:

```suggestion
EKS is the one resource type that may need extra setup on your side. Two things to be aware of:
```

## Verify it is working

Once deployed, three places tell you Kosli Capture is doing its job:

1. **CloudWatch Logs** for the Lambda function — each run logs the scan start, the resources it discovered vs. put in scope, the env names it derived, any gaps, and any errors. Bump verbosity at runtime (no rebuild) with the CloudFormation parameter `DebugLevel=debug` or `DebugLevel=trace`. `trace` also passes `--debug` into the underlying `kosli` CLI. The API token is never logged.
2. **[Kosli Environments](https://app.kosli.com)** — new environments appear under **Environments** with names from your template. Each snapshot shows the running artifacts and their fingerprints, exactly as if you had reported them by hand.
3. **[`kosli list snapshots`](/client_reference/kosli_list_snapshots)** and [`kosli get snapshot`](/client_reference/kosli_get_snapshot) — query environments from the CLI.

## Good to know

> Behavior worth being explicit about, none of it blocking:

Reviewer: "none of it blocking" is PR-review vocabulary — "blocking" describes review states, not product behavior a customer experiences:

```suggestion
A few behaviors to be aware of:
```

- **One region per deploy by default.** List `scope.regions` to fan out across regions, and add `{region}` to your naming template so each region gets its own env.
- **Kosli Capture excludes itself** from Lambda discovery, so its own Lambda never shows up in your Kosli environments.

> - **EventBridge triggers are not enabled by default.** The two-minute schedule is the backbone — EKS pods do not emit AWS events, so schedule-driven scanning is the common denominator. You can additionally invoke the same function from EventBridge rules (ECS task state change, CloudTrail `CreateCluster` / `CreateBucket`) for lower-latency updates on the AWS-native types; every run re-derives truth, so no payload handling is needed on the trigger.

Reviewer: "The two-minute schedule is the backbone", "the common denominator", and "every run re-derives truth" are architecture-decision-record language. Customer-voiced version of the same content:

```suggestion
- **EventBridge triggers are not enabled by default.** Kosli Capture relies on its two-minute schedule because it covers every resource type — EKS pod changes do not emit AWS events. For lower-latency updates on ECS, Lambda, and S3, you can additionally invoke the same function from EventBridge rules (for example ECS task state change, or CloudTrail `CreateCluster` / `CreateBucket`); each run rescans everything, so the rules need no payload handling.
```

- **Scaling events are off by default.** Kosli does not create a new snapshot for a change in replica count only — the workload's provenance has not changed. Turn on `include_scaling: true` (globally or per type) if you want scaling to produce new snapshots.
- **Renaming an env-name tag creates a new env, not a rename.** Kosli Capture cannot see the previous value; it just derives the new name and creates the env on first sight. The old env remains until you archive it.

## Constraints

- **Deploy in a supported region and keep the function `x86_64`.** The Kosli CLI layer is published only in the nine regions listed under [Prerequisites](#prerequisites), and the function architecture has to match the layer's.

> - **Private-only EKS endpoints require in-VPC deployment.** See the [EKS caveats](#eks-caveats-the-honest-part).

Reviewer: If the earlier suggestion to rename the heading to "## EKS caveats" (line 214) is accepted, this anchor needs updating to match:

```suggestion
- **Private-only EKS endpoints require in-VPC deployment.** See the [EKS caveats](#eks-caveats).
```
